inst/include/utils.h

```diff
@@ -1,5 +1,75 @@
 #include <torch.h>
 
+#include <stack>
+#include <mutex>
+#include <condition_variable>
+#include <future>
+#include <functional>
+#include <utility>
+#include <type_traits>
+
+template <typename F>
+class ScopeGuard {
+ public:
+  explicit ScopeGuard(F&& f) noexcept : f_(std::forward<F>(f)) {}
+  ~ScopeGuard() noexcept { f_(); }
+ private:
+  typename std::decay<F>::type f_;
+};
+
+// TODO:
+// create a factory function for ScopeGuard that disallows the returned
+// ScopeGuard object to be discarded immediately (but we need to do so in a
+// portable manner, by using the [[nodiscard]] attribute if possible, or
```

Comment:
Having `[[nodiscard]]` here is actually quite important because the ScopeGuard object is supposed to be destroyed when out of scope rather than immediately, because if someone were to mistakenly have

```
makeScopeGuard(fn);
```

or

```
ScopeGuard(fn);
```

instead, then the `ScopeGuard` object gets destroyed after the `;`, which would make it not work as expected.

The `[[nodiscard]]` attribute is only standardized in C++-17 though. I'm not sure whether there is a portable solution here if we want to stick with C++-11.